### PR TITLE
Img protocol

### PR DESCRIPTION
**Pull request by LeslieJRP**
#### What problem is this solving?

Use the image protocol to display personalized images.

#### How to test it?

To test this we need to have all three apps (store-image, image-protocol and image-protocol-example) linked to our workspace and have saved URLs for a specific customer class and image id. 

[Workspace](https://imageprotocol--lreyes.myvtex.com/)

#### Screenshots or example usage:

Captura de pantalla 2022-05-31 a las 17 26 20

Captura de pantalla 2022-05-31 a las 17 27 30

#### Related to / Depends on

This change depends on the image-protocol app

#### How does this PR make you feel? [:link:](http://giphy.com/)

Happy!

### CODE REVIEW

**Comment by sheilagomes** [comment]
I've opened task [LOC-8460](https://vtex-dev.atlassian.net/browse/LOC-8460) for this one, but please @LeslieJRP , when you need translations again, open a JIRA task for us following the workflow in our #localization slack channel.

File: manifest.json [hunk 1]
```diff
@@ -11,13 +11,16 @@
     "docs": "0.x"
   },
   "dependencies": {
+    "vtex.styleguide": "9.x",
```
Comment (tomymehdi) [inline manifest.json]: do we need to add this one?

File: messages/context.json [hunk 2]
```diff
@@ -1,5 +1,6 @@
 {
   "admin/editor.store-image.title": "Image",
+  "admin/editor.store-image.imageProtocolId.title": "Protocol Image id",
```
Comment (tomymehdi) [inline messages/context.json]: Image Protocl Id

File: messages/context.json [hunk 3]
```diff
@@ -8,6 +9,7 @@
   "admin/editor.image-list.description": "Show a list of images",
   "admin/editor.image-list.images.title": "Images",
   "admin/editor.image-list.images.image.title": "Image",
+  "admin/editor.image-list.images.imageProtocolId.title": "Protocol Image Id",
```
Comment (tomymehdi) [inline messages/context.json]: Image Protocl Id

File: messages/en.json [hunk 4]
```diff
@@ -1,5 +1,6 @@
 {
   "admin/editor.store-image.title": "Image",
+  "admin/editor.store-image.imageProtocolId.title": "Image id",
```
Comment (tomymehdi) [inline messages/en.json]: Image Protocl Id

File: messages/en.json [hunk 5]
```diff
@@ -8,6 +9,7 @@
   "admin/editor.image-list.description": "Show a list of images",
   "admin/editor.image-list.images.title": "Images",
   "admin/editor.image-list.images.image.title": "Image",
+  "admin/editor.image-list.images.imageProtocolId.title": "Id de la Imagen de Protocolo",
```
Comment (tomymehdi) [inline messages/en.json]: Image Protocl Id

File: messages/es.json [hunk 6]
```diff
@@ -1,5 +1,6 @@
 {
   "admin/editor.store-image.title": "Imagen",
+  "admin/editor.store-image.imageProtocolId.title": "Protocol Image id",
```
Comment (tomymehdi) [inline messages/es.json]: Image Protocl Id

File: messages/es.json [hunk 7]
```diff
@@ -8,6 +9,7 @@
   "admin/editor.image-list.description": "Mostrar una lista de imágenes",
   "admin/editor.image-list.images.title": "Imágenes",
   "admin/editor.image-list.images.image.title": "Imagen",
+  "admin/editor.image-list.images.imageProtocolId.title": "Protocol Image Id",
```
Comment (tomymehdi) [inline messages/es.json]: Image Protocl Id

File: messages/pt.json [hunk 8]
```diff
@@ -1,5 +1,6 @@
 {
   "admin/editor.store-image.title": "Imagem",
+  "admin/editor.store-image.imageProtocolId.title": "Protocol Image id",
```
Comment (tomymehdi) [inline messages/pt.json]: Image Protocl Id

File: messages/pt.json [hunk 9]
```diff
@@ -8,6 +9,7 @@
   "admin/editor.image-list.description": "Exiba uma lista de imagens",
   "admin/editor.image-list.images.title": "Imagens",
   "admin/editor.image-list.images.image.title": "Imagem",
+  "admin/editor.image-list.images.imageProtocolId.title": "Protocol Image Id",
```
Comment (tomymehdi) [inline messages/pt.json]: Image Protocl Id

File: react/Image.tsx [hunk 10]
```diff
@@ -7,14 +8,17 @@ import { useIntl, defineMessages } from 'react-intl'
 import { formatIOMessage } from 'vtex.native-types'
 import { Link } from 'vtex.render-runtime'
 import { usePixel } from 'vtex.pixel-manager'
-
 import type { ImageSchema } from './ImageTypes'
+import GET_ImgUrl from './graphql/getImgUrl.gql'
```
Comment (tomymehdi) [inline react/Image.tsx]: import GET_IMAGE_PROTOCOL_IMAGES from './graphql/getImageProtocolImages.gql'

File: react/Image.tsx [hunk 11]
```diff
@@ -115,26 +121,83 @@ function Image(props: ImageProps) {
...
-      sizes={sizes}
-      srcSet={srcSet}
-      src={typeof formattedSrc === 'string' ? formattedSrc : ''}
-      alt={typeof formattedAlt === 'string' ? formattedAlt : ''}
-      style={imageDimensions}
-      ref={imageRef}
-      className={handles.imageElement}
-      {...(preload
-        ? {
+  const { loading2, session, error2 } = useRenderSession()
```
Comment (tomymehdi) [inline react/Image.tsx]: Only execute useRenderSession if imageProtocolId is different from ''

Comment (tomymehdi) [reply react/Image.tsx]: To avoid executing this query if not needed and degrading performance of all loaded Images components

**Review by tomymehdi — CHANGES_REQUESTED** [review]

File: docs/README.md [hunk 12]
```diff
@@ -102,6 +102,61 @@ The block still doesn't have CSS Handles for its specific customization.
 All CSS Handles available for the Image block are the ones available for the `slider-layout` block. Take a look at the Customization section in the [**Slider Layout documentation**](https://vtex.io/docs/app/vtex.slider-layout).
 Note that the `image-slider` uses our `vtex.slider-layout` app, so all the CSS namespaces defined by it are also available for `image-slider`. Take a look at [Slider-Layout](https://vtex.io/docs/app/vtex.slider-layout).
 
+## Image Protocol Implementation
```
Comment (mariana-caetano) [inline docs/README.md]:
```suggestion
## Implementing Image Protocol 
```

File: docs/README.md [hunk 13]
```diff
@@ -102,6 +102,61 @@ The block still doesn't have CSS Handles for its specific customization.
...
+
```
Comment (mariana-caetano) [inline docs/README.md:115]:
```suggestion
The Image Protocol is an app that displays personalized images to use alongside the Store Image app in your store's images. Follow the steps below to implement the Image Protocol in your code.
```

File: docs/README.md [hunk 14]
```diff
@@ -102,6 +102,61 @@ The block still doesn't have CSS Handles for its specific customization.
...
+In the `manifest.json` we added Session Client app and also the Image Protocol app in the dependency list:
```
Comment (mariana-caetano) [inline docs/README.md]:
```suggestion
1. In your theme's `manifest.json`, add the Session Client app and also the Image Protocol app as a dependency:
```

File: docs/README.md [hunk 15]
```diff
@@ -102,6 +102,61 @@ The block still doesn't have CSS Handles for its specific customization.
...
+
+```json
+ "dependencies": {
+    "vtex.session-client": "1.x",
+    "vtex.image-protocol": "0.x"
+  }
+```
+The interface `ImageTypes.ts` is modified to have the isMobile and imageProtocolId props:
```
Comment (mariana-caetano) [inline docs/README.md:127]: @LeslieJRP, I didn't understand this part. Could you help me by answering the following questions?
1. Should the developer add the `ImageTypes.ts` on their code?
2. Is this an example of how should it look? 
3. Is this a step-by-step that the user should follow to modify the `ImageTypes.ts`?
Thanks!

File: docs/README.md [hunk 16]
```diff
@@ -102,6 +102,61 @@ The block still doesn't have CSS Handles for its specific customization.
...
+
+```ts
+export interface ImageSchema {
+  isMobile?: boolean
+  imageProtocolId? : string
+  ...
+}
+```
+To be able to personalize a specific image the props defined in the interface are also exposed in the Image component:
```
Comment (mariana-caetano) [inline docs/README.md]:
```suggestion
2. To personalize a specific image, the props defined in the interface must be exposed in the Image component as the example below: 
```

File: docs/README.md [hunk 17]
```diff
@@ -102,6 +102,61 @@ The block still doesn't have CSS Handles for its specific customization.
...
+
+```tsx
+const{
+  isMobile=false,
+  imageProtocolId='',
+  ...
+}=props
+``` 
+
+>**NOTE**: The prop imageProtocolId can be modified from the Site Editor.
```
Comment (mariana-caetano) [inline docs/README.md]:
```suggestion
> ⚠️
> 
> The prop `imageProtocolId` can be modified via Site Editor.
```

Comment (mariana-caetano) [reply docs/README.md]: How can we modify the prop via Site Editor?

File: docs/README.md [hunk 18]
```diff
@@ -102,6 +102,61 @@ The block still doesn't have CSS Handles for its specific customization.
...
+
+To implement the URLs from the image protocol example we have created inside the `react/graphql` the `getImgUrl.gql` that will use the resolver in the `image-protocol` app.
```
Comment (mariana-caetano) [inline docs/README.md:149]:
```suggestion
3. Implement the URLs from the image protocol by creating inside the `react/graphql` the `getImgUrl.gql` that will use the resolver in the `image-protocol` app. See the example below:
```

File: docs/README.md [hunk 19]
```diff
@@ -102,6 +102,61 @@ The block still doesn't have CSS Handles for its specific customization.
...
+```graphql
+query getImage($userId: String!, $imageProtocolId: String!) {
+    getImage(userId: $userId, imageProtocolId: $imageProtocolId) @context(provider: "vtex.image-protocol"){
+    url,
+    urlMobile
+  }
+}
+```
+
+In order to get the user Id we use the useRenderSession and the SessionSuccess from the vtex.session-client package in the `Image.tsx`. Also to use the query created inside the graphq folder we also import it as well as the useQuery.
```
Comment (mariana-caetano) [inline docs/README.md]:
```suggestion
4. Get the user Id using the `useRenderSession` and the `SessionSuccess` from the `vtex.session-client `package in the `Image.tsx`. Also, to use the query created inside the GraphQL folder, import the query as well as the `useQuery`.
```

File: docs/README.md [hunk 20]
```diff
@@ -102,6 +102,61 @@ The block still doesn't have CSS Handles for its specific customization.
...
+
+```tsx
+import { useQuery } from 'react-apollo'
+import GET_ImgUrl from './graphql/getImgUrl.gql'
+
+import { SessionSuccess, useRenderSession } from 'vtex.session-client'
+```
+
+Now as the Image component has the imageProtocolId prop if it has an Id and we can get the user Id we can use the query defined in `getImgUrl.gql` that receives these two variables and is skipped if no user id is provided.
```
Comment (mariana-caetano) [inline docs/README.md]:
```suggestion
> ⚠️
> 
> As the Image component has the `imageProtocolId` prop if it has an Id, you can get the user Id and use the query defined in `getImgUrl.gql` that receives these two variables. If no user id is provided, skip this suggestion.
```

Comment (mariana-caetano) [reply docs/README.md]: Could you explain more about this part?

File: docs/README.md [hunk 21]
```diff
@@ -102,6 +102,61 @@ The block still doesn't have CSS Handles for its specific customization.
...
+
+We set the src for this imgElement depending on the response of the query and also if this is for mobile or not. Other image elements that don't have the image protocol Id or the result of the query has the URLs as null we then set the default URLs.
```
Comment (mariana-caetano) [inline docs/README.md:174]: These two sentences could be a step-by-step

**Review by mariana-caetano — CHANGES_REQUESTED** [review]
Hi @LeslieJRP. Thanks for your contribution! I reviewed the README.md and made some suggestions and questions.

File: docs/README.md (hunk 17 above)
Comment (LeslieJRP) [reply docs/README.md]: If you go to the Site Editor and check the List of images, selecting any of the images you will see the field 'Image Protocol Id'
Captura de pantalla 2022-06-10 a las 12 10 56

File: docs/README.md (hunk 21 above)
Comment (LeslieJRP) [reply docs/README.md:174]: It's not a step by step because it's done in the code, I explained here how it works

File: messages/pt.json [hunk 22]
```diff
@@ -1,5 +1,6 @@
 {
   "admin/editor.store-image.title": "Imagem",
+  "admin/editor.store-image.imageProtocolId.title": "ID do protocolo de imagem",
```
Comment (sheilagomes) [inline messages/pt.json]:
```suggestion
  "admin/editor.store-image.imageProtocolId.title": "ID do protocolo da imagem",
```

File: messages/pt.json [hunk 23]
```diff
@@ -8,6 +9,7 @@
   "admin/editor.image-list.description": "Exiba uma lista de imagens",
   "admin/editor.image-list.images.title": "Imagens",
   "admin/editor.image-list.images.image.title": "Imagem",
+  "admin/editor.image-list.images.imageProtocolId.title": "ID do protocolo de imagem",
```
Comment (sheilagomes) [inline messages/pt.json:14]:
```suggestion
  "admin/editor.image-list.images.imageProtocolId.title": "ID do protocolo da imagem",
```

File: messages/es.json [hunk 24]
```diff
@@ -1,5 +1,6 @@
 {
   "admin/editor.store-image.title": "Imagen",
+  "admin/editor.store-image.imageProtocolId.title": "ID de protocolo de imagen",
```
Comment (sheilagomes) [inline messages/es.json]:
```suggestion
  "admin/editor.store-image.imageProtocolId.title": "ID del protocolo de la imagen",
```

File: messages/es.json [hunk 25]
```diff
@@ -8,6 +9,7 @@
   "admin/editor.image-list.description": "Mostrar una lista de imágenes",
   "admin/editor.image-list.images.title": "Imágenes",
   "admin/editor.image-list.images.image.title": "Imagen",
+  "admin/editor.image-list.images.imageProtocolId.title": "ID de protocolo de imagen",
```
Comment (sheilagomes) [inline messages/es.json:14]:
```suggestion
  "admin/editor.image-list.images.imageProtocolId.title": "ID del protocolo de la imagen",
```

**Comment by sheilagomes** [comment]
> I've opened task [LOC-8460](https://vtex-dev.atlassian.net/browse/LOC-8460) for this one, but please @LeslieJRP , when you need translations again, open a JIRA task for us following the workflow in our #localization slack channel.

Done! Translations are suggested in the code.

File: docs/README.md (hunk 15 above)
Comment (LeslieJRP) [reply docs/README.md:127]: Hi @Mariana-Caetano 
1. The developer doesn't need to add anything in the ImageTypes.ts, this is already done
2. The code should be like this, no need to modify anything
3. No, the user doesn't need to modify anything in the code, to modify the props the user can use the json file or the site editor

Hope this helps, let me know if you need any other clarification.
Thanks!

**Review by LeslieJRP — COMMENTED** [review]
all suggestions were reviewed

**Review by LeslieJRP — COMMENTED** [review]
all changes were done

**Review by LeslieJRP — COMMENTED** [review]
all the changes were checked

**Review by LeslieJRP — COMMENTED** [review]
all suggestions viewed

**Review by LeslieJRP — COMMENTED** [review]
checked

**Review by LeslieJRP — COMMENTED** [review]
checked

**Review by sheilagomes — APPROVED** [review]

**Pull request closed without merging**